### PR TITLE
When indexed-repeat is used within a repeat, 1st, 2nd, 4th, and 6th a…

### LINKED ISSUE #484: When indexed-repeat is used within a repeat the first argument nodeset points to single node using relative path
#### Software and hardware versions 
2.2.1

#### Problem description
The first argument of indexed-repeat() calls is relative when called from inside a repeat and referring to a question in that repeat. It should be absolute.

A `../relative/path/to/question` always returns a single node (in ODK XForms) and according to [the spec](https://getodk.github.io/xforms-spec/#fn:indexed-repeat) it should refer to all nodes with that question name, i.e. the nodeset returned by `/absolute/path/to/question`.

#### Steps to reproduce the problem

_A. using regular calculation with relative index:_

type | name | label | calculation
-- | -- | -- | --
begin_repeat | person | Person |  
integer | pos | | position(..) |
text | name | Enter name |  
text | prev_name | Name in previous repeat instance | indexed-repeat(${name}, ${person}, ${pos}-1)
end_repeat |   |   |  

outputs: 

```xml
<bind calculate="indexed-repeat( ../name ,  /data/person , ../pos -1)"  ... />
```

_B. using dynamic default:_

type | name | label | default
-- | -- | -- | --
begin_repeat | person | Person |  
text | name | Enter name |  
text | prev_name | Name in previous repeat instance | indexed-repeat(${name}, ${person}, position(..)-1)
end_repeat |   |   |  

outputs:

```xml
<setvalue ... value="indexed-repeat( ../name ,  /data/person , position(..)-1)" />
```

_C. using dreaded nested repeats:_

type | name | label | default
-- | -- | -- | --
begin_repeat | family | Family |  
integer | members_number | How many members in this family? |  
begin_repeat | person | Person |  
text | name | Enter name |  
text | prev_name | Non-sensible previous name in first family, 2nd person | indexed-repeat(${name}, ${family}, 1, ${person}, 2)
end_repeat |   |   |  
end_repeat |   |   |  

outputs:

```xml
<setvalue .. value="indexed-repeat( ../name ,  /data/family , 1,  ../../person , 2)" />
```

#### Expected behavior

A:

```xml
<bind calculate="indexed-repeat( /data/person/name ,  /data/person , ../pos - 1)" ... />
```

B: 
```xml
<setvalue ... value="indexed-repeat( /data/person/name ,  /data/person , position(..)-1)" />
```

C: 
```xml
<setvalue .. value="indexed-repeat( /data/family/person/name ,  /data/family , 1, /data/family/person , 2)" />
```

So the first argument of indexed-repeat() is an exception to the relative-path rule used to transform ${node} syntax (and the second argument too but that seems to already be the case).

Or actually since indexed-repeat() has either 3, 5, or 7 arguments, this exception applies to the 1st, 2nd, 4th, and 6th  arguments.

#### Other information 

Previously using indexed-repeat() inside a repeat was a bit of an edge case. Since the introduction of dynamic defaults it is becoming more common, as you could use the answer of a previous repeat question as the dynamic default of that same question in a new repeat instance (example B).

I'm guessing this issue was introduced in 0.12.0 when we changed to relative node references inside repeats.

### PR DESCRIPTION
**Pull request by gushil**
…rgument not using relative path.

Closes #
Fix #484 
#### Why is this the best possible solution? Were any other approaches considered?
The one that works and passed tests.
#### What are the regression risks?
All tests passed, including new ones.
#### Does this change require updates to documentation? If so, please file an issue [here](https://github.com/XLSForm/xlsform.github.io) and include the link below.
No
#### Before submitting this PR, please make sure you have:
- [x] included test cases for core behavior and edge cases in `tests_v1`
- [x] run `nosetests` and verified all tests pass
- [x] run `black pyxform` to format code
- [x] verified that any code or assets from external sources are properly credited in comments

### CODE REVIEW

**Review by MartijnR — COMMENTED** [review]
Thanks @gushil!

I checked the tests and this all looks good. I also double-checked the output when indexed-repeat is used in a constraint, relevant and required, and this is also correct.

I glanced at the code a little bit, and all the if-statements seemed unexpected to me. However, I have no understanding of the intricacies of how the current ${node} references are process when referring to nodes inside repeats from that repeat. So, I am hoping this makes immediate sense to @lognaturel...

**Comment by lognaturel** [comment]
> all the if-statements seemed unexpected to me.

Agreed, this looks tough to maintain and evolve. Perhaps it's the only reasonable solution but it would really help to have a narrative on what alternatives you considered and why you think this is the best/only approach. Otherwise the only way I can provide meaningful review is to solve the problem myself all over again. It appears that you are explicitly identifying all the different places that an `indexed-repeat` function could appear. Is there not a single place where that could be performed regardless of where `indexed-repeat` appeared? At the very least, this logic should be grouped in a function with a clear comment on its purpose.

**Comment by gushil** [comment]
@lognaturel @MartijnR 
Thanks for the reviews!

I agree with your comment about the code looks quite difficult to understand. I try to not moving the code too much, but it looks like it made it difficult to understand and too many conditional repetitions.

Basically, I want to add conditions related to this issue in addition to the existing relative path condition:

`
not (
    context["type"] == "calculate"
    and "indexed-repeat" in context["bind"]["calculate"]
)
`

But it looks like, some of the conditions related to this issue are covered in the existing condition, so I break the current condition and add conditions related to this issue.

Another problem I found is that I cannot tell if this name to be replaced is in indexed-repeat without checking all conditions (context["bind"]["calculate"] and context["default"]), so I made those checks.

I will try to make it simpler and add some comments to it.
I'm open to any suggestions.

Thanks

**Comment by gushil** [comment]
Hi, @lognaturel @MartijnR 

Code has been refactored and all additional tests passed.

Hopefully, the current implementation is simpler and easy to understand. I've added some comments too.

Thanks.

File: pyxform/tests_v1/test_repeat.py [hunk 1]
```diff
@@ -458,3 +458,133 @@ def test_choice_from_previous_repeat_answers_in_nested_repeat_uses_current(self)
...
+    def test_multiple_indexed_repeat_in_epression_dreaded_nested_repeat_relative_path_exception(
+        self,
+    ):
+        """Test relative path exception (absolute path) in multiple indexed-repeat() inside an expression using dreaded nested repeat."""
+        self.assertPyxformXform(
+            name="data",
+            title="dreaded nested repeat indexed-repeat 1st, 2nd, 4th, and 6th parameters is using absolute path",
+            md="""
+                | survey  |                |                |                                  |                                                                                                                 |
+                |         | type           | name           | label                            | default                                                                                                         |
```
Comment (MartijnR) [inline pyxform/tests_v1/test_repeat.py:553]: Change "default" to "required"? (Though the expression makes little sense in that case - not returning a boolean -, it doesn't really matter for these tests, I think).

**Comment by lognaturel** [comment]
Thanks for the good rework, @gushil. I'm comfortable with it now.

One case I've found that isn't handled is if the `indexed-repeat` call is nested in another call and there's a relative reference parameter after. For example, consider `concat(indexed-repeat(${name}, ${family}, 1, ${person}, 2), ${name})` in a nested repeat. The output should be `concat(indexed-repeat( /data/family/person/name ,  /data/family , 1,  /data/family/person , 2), ../name )` but is `concat(indexed-repeat( /data/family/person/name ,  /data/family , 1,  /data/family/person , 2),  /data/family/person/name )`.

Putting the `${name}` as the first parameter works: `concat(${name}, indexed-repeat(${name}, ${family}, 1, ${person}, 2))` correctly produces `concat( ../name , indexed-repeat( /data/family/person/name ,  /data/family , 1,  /data/family/person , 2))`.

File: pyxform/tests_v1/test_repeat.py [hunk 2]
```diff
@@ -458,3 +458,133 @@ def test_choice_from_previous_repeat_answers_in_nested_repeat_uses_current(self)
...
+            xml__contains=[
+                """<setvalue event="odk-instance-first-load odk-new-repeat" ref="/data/person/prev_name" value="indexed-repeat( /data/person/name ,  /data/person , position(..)-1)"/>"""  # noqa pylint: disable=line-too-long
+            ],
+        )
+
+    def test_indexed_repeat_dreaded_nested_repeat_relative_path_exception(self):
+        """Test relative path exception (absolute path) in indexed-repeat() using dreaded nested repeat."""
+        self.assertPyxformXform(
+            name="data",
+            title="dreaded nested repeat indexed-repeat 1st, 2nd, 4th, and 6th parameters is using absolute path",
```
[the remark this answers is not in the document]
Comment (gushil) [reply pyxform/tests_v1/test_repeat.py:503]: Will be removed in next commit

File: pyxform/tests_v1/test_repeat.py [hunk 3]
```diff
@@ -458,3 +458,133 @@ def test_choice_from_previous_repeat_answers_in_nested_repeat_uses_current(self)
...
+    def test_indexed_repeat_math_epression_with_double_variable_in_dreaded_nested_repeat_relative_path_exception(
+        self,
+    ):
+        """Test relative path exception (absolute path) in indexed-repeat() with math expression and double variable using dreaded nested repeat."""
+        self.assertPyxformXform(
+            name="data",
+            title="dreaded nested repeat indexed-repeat 1st, 2nd, 4th, and 6th parameters is using absolute path",
+            md="""
+                | survey  |                |                |                                  |                                                             |
+                |         | type           | name           | label                            | default                                                     |
```
[the remark this answers is not in the document]
Comment (gushil) [reply pyxform/tests_v1/test_repeat.py:577]: Thanks. Will be fixed in next commit.

File: pyxform/tests_v1/test_repeat.py [hunk 4]
```diff
@@ -458,3 +458,133 @@ def test_choice_from_previous_repeat_answers_in_nested_repeat_uses_current(self)
...
+    def test_indexed_repeat_math_epression_dreaded_nested_repeat_relative_path_exception(
+        self,
+    ):
+        """Test relative path exception (absolute path) in indexed-repeat() with math expression using dreaded nested repeat."""
+        self.assertPyxformXform(
+            name="data",
+            title="dreaded nested repeat indexed-repeat 1st, 2nd, 4th, and 6th parameters is using absolute path",
+            md="""
+                | survey  |                |                |                                  |                                                        |
+                |         | type           | name           | label                            | default                                                |
```
[the remark this answers is not in the document]
Comment (gushil) [reply pyxform/tests_v1/test_repeat.py:529]: Got it. WIll be fixed in the next commit.

File: pyxform/tests_v1/test_repeat.py (hunk 1 above)
Comment (gushil) [reply pyxform/tests_v1/test_repeat.py:553]: Ok. Let's try using "required". WIll be changed in the next commit.

**Comment by gushil** [comment]
Hi @lognaturel @MartijnR 

Addressed @MartijnR reviews and fixed the code (and add a test too) to fix the bug that @lognaturel  mentioned in the feedback.

Thanks.

**Review by lognaturel — APPROVED** [review]
👍 

In general, doing argument parsing with regex (as opposed to a parser) scares me a bit but based on the tests, I think you've pulled it off. I think there are probably still cases like commas in quotes that could break but I think it's ok to wait and see whether it's something users actually end up needing.
